workflow/controller/workflowpod.go

```diff
@@ -389,6 +409,10 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
 		return nil, err
 	}
 
+	if err := woc.scheduleOnDifferentHost(node, pod); err != nil {
```

Comment:
I found this place to be the best for injecting the `affinity`, after the constraints in `addSchedulingConstraints()` and `podSpecPatch` have been already applied.